lib/src/source/hosted.dart

```diff
@@ -374,20 +425,22 @@ class HostedSource extends CachedSource {
     final List<_VersionInfo> result;
     try {
       // TODO(sigurdm): Implement cancellation of requests. This probably
-      // requires resolution of: https://github.com/dart-lang/sdk/issues/22265.
-      bodyText = await withAuthenticatedClient(
-        cache,
-        Uri.parse(hostedUrl),
-        (client) => client.read(url, headers: pubApiHeaders),
-      );
+      // requires resolution of: https://github.com/dart-lang/http/issues/424.
+      bodyText = await withAuthenticatedClient(cache, Uri.parse(hostedUrl),
+          (client) async {
+        return await retryForHttp(
+            'fetching versions for "$packageName" from "$url"', () async {
+          return await client.read(url,
+              headers: HostedSource.httpRequestHeadersFor(url));
+        });
+      });
       final decoded = jsonDecode(bodyText);
       if (decoded is! Map<String, dynamic>) {
         throw FormatException('version listing must be a mapping');
       }
       body = decoded;
       result = _versionInfoFromPackageListing(body, ref, url, cache);
     } on Exception catch (error, stackTrace) {
-      final packageName = _asDescription(ref.description).packageName;
```

Comment:
Will removing `_asDescription` cause any side effects? I moved the `packageName` definition above since I use it in the try block. Then I noticed there's already a type check on line 415 which makes that call unnecessary:
https://github.com/dart-lang/pub/blob/e6ae9474bcfe23369911c2b36e8f0453fab872d9/lib/src/source/hosted.dart#L411-L421

Reply:
I think that can be removed. That looks much better!